runtime/common/ArgumentConversion.cpp

```diff
@@ -0,0 +1,360 @@
+  return builder.create<arith::ConstantIntOp>(builder.getUnknownLoc(), v, bits);
+}
+
+static Value genConstant(OpBuilder &builder, bool v) {
+  return genIntegerConstant(builder, v, 1);
+}
+static Value genConstant(OpBuilder &builder, char v) {
+  return genIntegerConstant(builder, v, 8);
+}
+static Value genConstant(OpBuilder &builder, std::int16_t v) {
+  return genIntegerConstant(builder, v, 16);
+}
+static Value genConstant(OpBuilder &builder, std::int32_t v) {
+  return genIntegerConstant(builder, v, 32);
+}
+static Value genConstant(OpBuilder &builder, std::int64_t v) {
+  return genIntegerConstant(builder, v, 64);
+}
+
+static Value genConstant(OpBuilder &builder, float v) {
+  return builder.create<arith::ConstantFloatOp>(
+      builder.getUnknownLoc(), APFloat{v}, builder.getF32Type());
+}
+static Value genConstant(OpBuilder &builder, double v) {
+  return builder.create<arith::ConstantFloatOp>(
+      builder.getUnknownLoc(), APFloat{v}, builder.getF64Type());
+}
+
+template <typename A>
+Value genComplexConstant(OpBuilder &builder, const std::complex<A> &v,
+                         FloatType fTy) {
+  auto rePart = builder.getFloatAttr(fTy, APFloat{v.real()});
+  auto imPart = builder.getFloatAttr(fTy, APFloat{v.imag()});
+  auto complexAttr = builder.getArrayAttr({rePart, imPart});
+  auto loc = builder.getUnknownLoc();
+  auto ty = ComplexType::get(fTy);
+  return builder.create<complex::ConstantOp>(loc, ty, complexAttr).getResult();
+}
+
+static Value genConstant(OpBuilder &builder, std::complex<float> v) {
+  return genComplexConstant(builder, v, builder.getF32Type());
+}
+static Value genConstant(OpBuilder &builder, std::complex<double> v) {
+  return genComplexConstant(builder, v, builder.getF64Type());
+}
+static Value genConstant(OpBuilder &builder, FloatType fltTy, long double *v) {
+  return builder.create<arith::ConstantFloatOp>(
+      builder.getUnknownLoc(),
+      APFloat{fltTy.getFloatSemantics(), std::to_string(*v)}, fltTy);
+}
+
+static Value genConstant(OpBuilder &builder, const std::string &v,
+                         ModuleOp substMod) {
+  auto loc = builder.getUnknownLoc();
+  cudaq::IRBuilder irBuilder(builder);
+  auto cString = irBuilder.genCStringLiteralAppendNul(loc, substMod, v);
+  auto addr = builder.create<cudaq::cc::AddressOfOp>(
+      loc, cudaq::cc::PointerType::get(cString.getType()), cString.getName());
+  auto i8PtrTy = cudaq::cc::PointerType::get(builder.getI8Type());
+  auto cast = builder.create<cudaq::cc::CastOp>(loc, i8PtrTy, addr);
+  auto size = builder.create<arith::ConstantIntOp>(loc, v.size(), 64);
+  auto chSpanTy = cudaq::cc::CharspanType::get(builder.getContext());
+  return builder.create<cudaq::cc::StdvecInitOp>(loc, chSpanTy, cast, size);
+}
+
+static Value genConstant(OpBuilder &builder, const cudaq::state *v,
+                         ModuleOp substMod) {
+  // TODO: do we materialize the data here or just add a symbolic reference into
+  // the unknown?
```

Comment:
Should we have a `TODO` exception here so that it might be easier to spot it in the future?